src/libcollections/vec.rs

```diff
@@ -1039,16 +1039,21 @@ impl<T> Vec<T> {
     #[inline]
     #[stable(feature = "append", since = "1.4.0")]
     pub fn append(&mut self, other: &mut Self) {
-        self.reserve(other.len());
-        let len = self.len();
         unsafe {
-            ptr::copy_nonoverlapping(other.as_ptr(), self.get_unchecked_mut(len), other.len());
+            self.append_slice(other.as_slice());
+            other.set_len(0);
         }
+    }
 
+    /// Appends a slice to `Self`.
+    ///
+    /// Does not drop elements from slice, must be handled by caller.
+    #[inline]
+    unsafe fn append_slice(&mut self, other: &[T]) {
```

Comment:
We use `other` here to transfer ownership of the elements, I'd be more comfortable with passing them as `other: *mut [T]`, and it could also be even more clear in the comment  that elements are moved from other; especially since elements are not dropped after actual use of the method. (No strong opinion on which `*const` or `*mut` to use.)